### PR TITLE
bump controller-runtime, k8s/golang/workflow-controller version

### PR DESCRIPTION
**Pull request by ccfishk**
Signed-off-by: jhu02 <jiminh@jiminhs-MacBook-Pro.local>

Fixes https://github.com/keikoproj/addon-manager/issues/93

### CODE REVIEW

File: controllers/addon_controller.go [hunk 1]
```diff
@@ -173,71 +175,77 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
+		OwnerType:    &addonmgrv1alpha1.Addon{},
+	}); err != nil {
+		return nil, err
 	}
 
 	// Watch for changes to kubernetes Resources matching addon labels.
-	for _, resc := range resources {
-		gvk := resc.GetObjectKind().GroupVersionKind()
-		_, kind := gvk.ToAPIVersionAndKind()
+	if err := c.Watch(&source.Kind{Type: &addonmgrv1alpha1.Addon{}}, &handler.EnqueueRequestForObject{}); err != nil {
```
[the remark this answers is not in the document]
Comment (ccfishk) [reply controllers/addon_controller.go:204]: Because we want leverage the built-in library directly, do not need extra informers, which has already with the k8s client.

File: go.mod [hunk 2]
```diff
@@ -1,35 +1,83 @@
...
-	k8s.io/api v0.19.11
-	k8s.io/apiextensions-apiserver v0.19.11
-	k8s.io/apimachinery v0.19.11
-	k8s.io/client-go v0.19.11
-	k8s.io/klog/v2 v2.3.0 // indirect
-	k8s.io/kube-openapi v0.0.0-20200831175022-64514a1d5d59 // indirect
-	k8s.io/utils v0.0.0-20200821003339-5e75c0163111 // indirect
-	sigs.k8s.io/controller-runtime v0.6.5
+	k8s.io/api v0.21.5
+	k8s.io/apiextensions-apiserver v0.21.2
```
Comment (kevdowney) [inline go.mod:17]: Why is this one not `v0.21.5`?

**Comment by kevdowney** [comment]
We need to verify this runs tests, please post unit-tests results `make test`

File: go.mod (hunk 2 above)
Comment (ccfishk) [reply go.mod:17]: this is compitible with all the dependencies. 
there is no difference between 21.2 and 21.5

File: main.go [hunk 3]
```diff
@@ -18,19 +18,20 @@ import (
...
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/log/zap"
 
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 	addonmgrv1alpha1 "github.com/keikoproj/addon-manager/api/v1alpha1"
 	"github.com/keikoproj/addon-manager/controllers"
+	"github.com/keikoproj/addon-manager/pkg/common"
 	"github.com/keikoproj/addon-manager/pkg/version"
-	// +kubebuilder:scaffold:imports
```
Comment (kevdowney) [inline main.go:29]: This is required I think.

File: main.go [hunk 4]
```diff
@@ -43,28 +44,31 @@ func init() {
 		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
 	flag.BoolVar(&debug, "debug", false, "Debug logging")
 	flag.Parse()
-
-	_ = addonmgrv1alpha1.AddToScheme(scheme)
-	// +kubebuilder:scaffold:scheme
```
Comment (kevdowney) [inline main.go:48]: This is required by kubebuilder

File: main.go [hunk 5]
```diff
@@ -43,28 +44,31 @@ func init() {
...
+		LeaderElectionID:      "addonmgr.keikoproj.io",
+		ClientDisableCacheFor: nonCached,
 	})
 	if err != nil {
 		setupLog.Error(err, "unable to start manager")
 		os.Exit(1)
 	}
 
-	err = controllers.NewAddonReconciler(mgr, ctrl.Log.WithName("controllers").WithName("Addon")).SetupWithManager(mgr)
+	_, err = controllers.New(mgr)
```
Comment (kevdowney) [inline main.go:71]: Are we missing logger ?
```
ctrl.Log.WithName("controllers").WithName("Addon"))
```

File: controllers/addon_controller.go [hunk 6]
```diff
@@ -33,27 +33,30 @@ import (
...
 	addonmgrv1alpha1 "github.com/keikoproj/addon-manager/api/v1alpha1"
 	"github.com/keikoproj/addon-manager/pkg/addon"
 	"github.com/keikoproj/addon-manager/pkg/common"
 	"github.com/keikoproj/addon-manager/pkg/workflows"
 )
 
-// addon ttl time
-const TTL = time.Duration(1) * time.Hour // 1 hour
+const (
+	controllerName = "addon_manager_controller"
```
Comment (kevdowney) [inline controllers/addon_controller.go:56]:
```suggestion
	controllerName = "addon-manager"
```

File: controllers/addon_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kevdowney) [reply controllers/addon_controller.go:204]: Since you made this change why do we need `resources` array now?
https://github.com/keikoproj/addon-manager/blob/bf0af606319b0ad1489813d2283d32333996af4b/controllers/addon_controller.go#L63-L70

[the remark this answers is not in the document]
Comment (kevdowney) [reply controllers/addon_controller.go:204]: Also you're using empty value for the types, which has null values.
```
&appsv1.Deployment{}
```
as opposed to resources list version is same but has kind and version populated:
```
&appsv1.Deployment{TypeMeta: metav1.TypeMeta{Kind: "Deployment", APIVersion: "apps/v1"}}
```

File: api/v1alpha1/suite_test.go [hunk 7]
```diff
@@ -46,7 +47,7 @@ func TestAPIs(t *testing.T) {
 }
 
 var _ = BeforeSuite(func(done Done) {
-	logf.SetLogger(zap.LoggerTo(GinkgoWriter, true))
+	logf.SetLogger(log)
```
Comment (kevdowney) [inline api/v1alpha1/suite_test.go:50]: `logr` is an interface, do the logs still get output here during unit-tests?

File: Dockerfile [hunk 8]
```diff
@@ -1,12 +1,27 @@
...
+# Build
+RUN CGO_ENABLED=0 GOOS=linux GO111MODULE=on go build -a -o manager main.go
 
 # Use distroless as minimal base image to package the manager binary
 # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+FROM gcr.io/distroless/static:nonroot AS distroless
 WORKDIR /
-COPY ${SOURCE_BINARY} .
-EXPOSE 8080 8443
```
Comment (shailshah9) [inline Dockerfile:8]: We need this for kube-proxy right?

File: config/crd/patches/cainjection_in_addons.yaml [hunk 9]
```diff
@@ -1,6 +1,6 @@
 # The following patch adds a directive for certmanager to inject CA into the CRD
 # CRD conversion requires k8s 1.13 or later.
```
Comment (shailshah9) [inline config/crd/patches/cainjection_in_addons.yaml:2]:
```suggestion
# CRD conversion requires k8s 1.16 or later.
```

File: config/crd/patches/webhook_in_addons.yaml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # The following patch enables conversion webhook for CRD
 # CRD conversion requires k8s 1.13 or later.
```
Comment (shailshah9) [inline config/crd/patches/webhook_in_addons.yaml:2]:
```suggestion
# CRD conversion requires k8s 1.16 or later.
```

File: api/v1alpha1/suite_test.go (hunk 7 above)
Comment (ccfishk) [reply api/v1alpha1/suite_test.go:50]: Yes, we did.

File: controllers/addon_controller.go [hunk 11]
```diff
@@ -173,71 +173,77 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
+
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: r})
 	if err != nil {
-		log.Error(err, "Error adding InformerFactory to the Manager")
-		return err
+		return nil, err
+	}
+
+	// Watch workflows created by addon only in addon-manager-system namespace
+	if err := c.Watch(&source.Kind{Type: &wfv1.Workflow{}}, &handler.EnqueueRequestForOwner{
```
Comment (kevdowney) [inline controllers/addon_controller.go:194]: This cannot work, we must only watch Workflows in the controller namespace. Before we limited to the controller ns here:
```
nsInformers := dynamicinformer.NewFilteredDynamicSharedInformerFactory(r.dynClient, time.Minute*30, managedNS, nil)
	wfInf := nsInformers.ForResource(common.WorkflowGVR())
	bldr := ctrl.NewControllerManagedBy(mgr).
		For(&addonmgrv1alpha1.Addon{}).
		// Watch workflows created by addon only in addon-manager-system namespace
		Watches(&source.Informer{Informer: wfInf.Informer().(cache.Informer)}, &handler.EnqueueRequestForOwner{
			IsController: true,
			OwnerType:    &addonmgrv1alpha1.Addon{},
		})
```

**Review by kevdowney — CHANGES_REQUESTED** [review]
Please restrict Workflow watcher to the controller namespace.

File: main.go (hunk 4 above)
Comment (ccfishk) [reply main.go:48]: We already customized the scheme. do not really need previous  one any more

File: Dockerfile (hunk 8 above)
Comment (ccfishk) [reply Dockerfile:8]: Those are default port. Addon-mgr does not really use them. Yeah, good observation

File: config/crd/patches/webhook_in_addons.yaml (hunk 10 above)
Comment (ccfishk) [reply config/crd/patches/webhook_in_addons.yaml:2]: we actually goes to 1.17. but, the comments are not blocker take it.

File: controllers/addon_controller.go (hunk 11 above)
Comment (ccfishk) [reply controllers/addon_controller.go:194]: when checking with resources, we filter using labelselector which pass namespace and name which is aligned specifically to the addon resource(s), which is implemented within the objects module.
thus, we do not need informers anymore

File: config/rbac/role.yaml [hunk 12]
```diff
@@ -99,7 +99,24 @@ rules:
...
+- apiGroups:
+  - coordination.k8s.io
+  resources:
+  - leases
+  verbs:
+  - '*'
+- apiGroups:
+  - argoproj.io
+  resources:
+  - workflows
```
Comment (kevdowney) [inline config/rbac/role.yaml:111]: This is required b/c you made the watcher change but this is not desired. We have previous fix to remove clusterrole access to watch Workflows due to memory issues. See #73

File: controllers/addon_controller.go (hunk 11 above)
Comment (kevdowney) [reply controllers/addon_controller.go:194]: We have previously fixed issue for memory consumption due to Watch (using the builder) see #73, which may be specifically an issue w/ CacheInformers but we also don't want to watch Workflows across all cluster, the design of Addon Manager at least for now is to run workflows only in `addon-manager-system` namespace so setting up a Watch for cluster is not efficent in the least and would be a regression in the worst case.

**Review by kevdowney — CHANGES_REQUESTED** [review]

File: main.go (hunk 3 above)
Comment (ccfishk) [reply main.go:29]: c71faae05d3b4bdee83d4be1b2e072c93c513265

File: controllers/addon_controller.go (hunk 11 above)
Comment (ccfishk) [reply controllers/addon_controller.go:198]: Only watch addon-manager-system namespace workflow
https://github.com/keikoproj/addon-manager/pull/96/commits/63d89c2dd79f8f770b3ae5a95f39f37e185d13af#diff-2df6a56a33260adb76f704ff7c6737be9f43c13fe523521ee76390ff2a9ca4b8R242

**Comment by ccfishk** [comment]
> We need to verify this runs tests, please post unit-tests results `make test`

jiminh@jiminhs-MacBook-Pro addon-manager % go test -timeout 5m -v ./test-bdd/...
=== RUN   TestE2e
Running Suite: Addon Type Suite
===============================
Random Seed: 1644305421
Will run 7 of 7 specs

W0207 23:30:21.460411   14502 warnings.go:70] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
Executing: /usr/local/bin/kubectl [apply -f /Users/jiminh/go/src/github.com/keikoproj/addon-manager/config/crd/bases/addonmgr.keikoproj.io_addons.yaml]W0207 23:30:23.466238   14502 warnings.go:70] apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
Executing: /usr/local/bin/kubectl [apply -f /Users/jiminh/go/src/github.com/keikoproj/addon-manager/config/crd/bases/argoproj_v1alpha1_workflows.yaml]•••••••
JUnit report was created: /Users/jiminh/go/src/github.com/keikoproj/addon-manager/test-bdd/junit.xml

Ran 7 of 7 Specs in 2.270 seconds
SUCCESS! -- 7 Passed | 0 Failed | 0 Pending | 0 Skipped
You're using deprecated Ginkgo functionality:
=============================================
Ginkgo 2.0 is under active development and will introduce (a small number of) breaking changes.
To learn more, view the migration guide at https://github.com/onsi/ginkgo/blob/v2/docs/MIGRATING_TO_V2.md
To comment, chime in at https://github.com/onsi/ginkgo/issues/711

  You are using a custom reporter.  Support for custom reporters will likely be removed in V2.  Most users were using them to generate junit or teamcity reports and this functionality will be merged into the core reporter.  In addition, Ginkgo 2.0 will support emitting a JSON-formatted report that users can then manipulate to generate custom reports.

  If this change will be impactful to you please leave a comment on https://github.com/onsi/ginkgo/issues/711
  Learn more at: https://github.com/onsi/ginkgo/blob/v2/docs/MIGRATING_TO_V2.md#removed-custom-reporters

To silence deprecations that can be silenced set the following environment variable:
  ACK_GINKGO_DEPRECATIONS=1.16.4

--- PASS: TestE2e (2.27s)
PASS

**Comment by ccfishk** [comment]
> Please restrict Workflow watcher to the controller namespace.

Yes.
63d89c2dd79f8f770b3ae5a95f39f37e185d13af

File: config/rbac/role.yaml (hunk 12 above)
Comment (ccfishk) [reply config/rbac/role.yaml:111]: This is resolved after watching only specific namespace

File: config/rbac/role.yaml [hunk 13]
```diff
@@ -99,7 +99,24 @@ rules:
...
   - patch
-
+- apiGroups:
+  - coordination.k8s.io
+  resources:
+  - leases
+  verbs:
+  - '*'
+- apiGroups:
+  - argoproj.io
```
Comment (kevdowney) [inline config/rbac/role.yaml:109]: This section needs to be removed from here.

Comment (ccfishk) [reply config/rbac/role.yaml:109]: No, we need this for docker image running other wise, serviceaccount:default could not get the leases. 
This is also the newer version requirement.

Comment (kevdowney) [reply config/rbac/role.yaml:109]: Not Leases, `argoproj.io/workflows`.

Comment (kevdowney) [reply config/rbac/role.yaml:109]: Remove the workflow rbac from clusterrole.

**Review by tekenstam — DISMISSED** [review]
LGTM. Just need to make sure comments from @kevdowney have been addressed.

File: config/rbac/role.yaml (hunk 13 above)
Comment (ccfishk) [reply config/rbac/role.yaml:109]: @kevdowney Do you mind share your thought why we need remove "leases", the controller would throw error "

leaderelection.go:325] error retrieving resource lock leases.coordination.k8s.io "addonmanager.keikoproj.io" is forbidden: User "system:serviceaccount:addon-manager-system:default" cannot get resource "leases"

Comment (kevdowney) [reply config/rbac/role.yaml:109]: Not the leases, remove the workflows rbac

Comment (kevdowney) [reply config/rbac/role.yaml:109]: I’m uncomfortable watching all the workflows in the entire cluster, there can thousands, we have a namespaced workflow-engine deployment and expect the rbac to match

Comment (ccfishk) [reply config/rbac/role.yaml]: 35dddd21c12b52cc2416d3133e41c3e4f51ab0aa removed

File: controllers/addon_controller.go [hunk 14]
```diff
@@ -158,6 +163,10 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
 		return reconcile.Result{Requeue: true}, nil
 	}
 
+	if addon.IsDuplicate(instance, r.versionCache) {
```
Comment (kevdowney) [inline controllers/addon_controller.go:166]: This will cause infinite reconcile, duplicate addons should be caught (ideally) by the addon_validator which does have a check for this. 
https://github.com/keikoproj/addon-manager/blob/c3ff3e626aa81547802fb051e009da6d6aed05db/pkg/addon/addon_validator.go#L60

Comment (ccfishk) [reply controllers/addon_controller.go:166]: This provides efficiency, reduces reconcile time - throw duplicate at earliest time. The addon_validator checks all other details though.

File: config/rbac/role.yaml [hunk 15]
```diff
@@ -99,7 +98,12 @@ rules:
   - get
   - list
   - patch
-
+- apiGroups:
+  - coordination.k8s.io
+  resources:
+  - leases
```
Comment (kevdowney) [inline config/rbac/role.yaml:104]: Just fyi. These rbac's are controlled by `// +kubebuilder:rbac:groups` pragma statements that generate this RBAC. So by adding it manually here it will probably be clobbered by next controller-gen execution; You'd need to add the proper rbac statement here:
https://github.com/keikoproj/addon-manager/blob/e1aac607f5312ca8d2640962a82cd84b97ab67b4/controllers/addon_controller.go#L111

Comment (ccfishk) [reply config/rbac/role.yaml:104]: Thanks for pointing this out. Currently, we are using controller-gen@v0.2.5, we should plan to bump to later version.

Comment (ccfishk) [reply config/rbac/role.yaml:104]: One of the points of adding rbac in yaml file is for user who is going to deploy the controller using pure yaml file. They might not run the step "make manifests."

I think we should have lease in both "marker comment" and yaml.

Comment (ccfishk) [reply config/rbac/role.yaml:104]: 4177242ed4f7af0c1e029985a9fe01adc7f3194e added

File: config/rbac/role.yaml (hunk 13 above)
Comment (ccfishk) [reply config/rbac/role.yaml]: We might have to adopt the "clusterrole" "argoproj.io" changes into the yaml file since we already restrict watching specific namespace workflows. This is also needed from controller-runtime

otherwise, controllers will fail with following error messages.
Failed to watch *v1alpha1.Workflow: failed to list *v1alpha1.Workflow: workflows.argoproj.io is forbidden: User "system:serviceaccount:addon-manager-system:default" cannot list resource "workflows" in API group "argoproj.io" at the cluster scope

File: config/rbac/role.yaml (hunk 15 above)
Comment (ccfishk) [reply config/rbac/role.yaml]: sort this out https://github.com/keikoproj/addon-manager/pull/96/commits/4177242ed4f7af0c1e029985a9fe01adc7f3194e

File: controllers/addon_controller.go [hunk 16]
```diff
@@ -98,24 +103,26 @@ func NewAddonReconciler(mgr manager.Manager, log logr.Logger) *AddonReconciler {
 // +kubebuilder:rbac:groups=addonmgr.keikoproj.io,resources=addons,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=addonmgr.keikoproj.io,resources=addons/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=argoproj.io,resources=workflows,namespace=system,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,namespace=system,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=core,resources=secrets,verbs=list
 // +kubebuilder:rbac:groups=rbac.authorization.k8s.io,resources=clusterroles;clusterrolebindings,verbs=get;list;patch;create
 // +kubebuilder:rbac:groups="",resources=namespaces;clusterroles;configmaps;events;pods;serviceaccounts;services,verbs=get;list;watch;create;update;patch
+// +kubebuilder:rbac:groups="argoproj.io",resources=workflows,verbs=get;list;watch;create;update;patch
```
Comment (kevdowney) [inline controllers/addon_controller.go:110]:
```suggestion

```
Remove this

File: controllers/addon_controller.go (hunk 14 above)
Comment (kevdowney) [reply controllers/addon_controller.go:168]: This short-circuits the validator which also sets `ValidationError` state on the Addon, so this change would infinitely loop without a resolution; we need to remove this code.

File: controllers/addon_controller.go (hunk 16 above)
Comment (ccfishk) [reply controllers/addon_controller.go:110]: @kevdowney The newer controller version requires all CR groups watched by clusterrole instead of role. 
Otherwise, controllers fails with error messages.
at least, this strategy is heavily used by all operators.

File: controllers/addon_controller.go (hunk 14 above)
Comment (ccfishk) [reply controllers/addon_controller.go:168]: I will re-test and see if can have a better solution.

Comment (ccfishk) [reply controllers/addon_controller.go:168]: Current issue is we always add new addon into cache w/o taking care of processAddon  errors. https://github.com/keikoproj/addon-manager/pull/96/files/e1aac607f5312ca8d2640962a82cd84b97ab67b4#diff-2df6a56a33260adb76f704ff7c6737be9f43c13fe523521ee76390ff2a9ca4b8R171-R174. 

Believing this cover some corner case.  But, if people purposely flush different addons (only name different, pkgName/Version is the same) into manager, we allow https://github.com/keikoproj/addon-manager/blob/master/pkg/addon/addon_version_cache.go#L63, which messed up the right Addon. 

This duplicateFilter at the first place try to cover this case.

Comment (ccfishk) [reply controllers/addon_controller.go]: Yes, this should be the graceful solution.
c7040b3703b64edfc9e86289b906b6163387394a

File: controllers/addon_controller.go (hunk 16 above)
Comment (kevdowney) [reply controllers/addon_controller.go:111]: Yes, but we have had issues this is a regression. This is not passable, it will consume cpu at least when there's 1000s.

File: main.go [hunk 17]
```diff
@@ -43,28 +45,31 @@ func init() {
...
-	_ = addonmgrv1alpha1.AddToScheme(scheme)
-	// +kubebuilder:scaffold:scheme
 }
 
 func main() {
 	ctrl.SetLogger(zap.New(zap.UseDevMode(debug)))
 
 	setupLog.Info(version.ToString())
 
+	nonCached := []client.Object{
```
Comment (figo) [inline main.go:55]: Why disabled cache for those objects? It is good to capture the reason with comment here.

File: Dockerfile [hunk 18]
```diff
@@ -1,12 +1,27 @@
-ARG SOURCE_BINARY=bin/manager
+FROM golang:1.17 as builder
```
[the remark this answers is not in the document]
Comment (ccfishk) [reply Dockerfile:1]: Yes, this is a good suggestion, always good to follow pattern. We can have alternative ticket address this comment. 

But this is defined intentionally this way, we want to test different dependencies on different layer, so builder is chosen as a k8s docker file pattern.

File: config/default/manager_auth_proxy_patch.yaml [hunk 19]
```diff
@@ -10,7 +10,7 @@ spec:
     spec:
       containers:
       - name: kube-rbac-proxy
-        image: gcr.io/kubebuilder/kube-rbac-proxy:v0.10.0
+        image: gcr.io/kubebuilder/kube-rbac-proxy:v0.8.0
```
[the remark this answers is not in the document]
Comment (ccfishk) [reply config/default/manager_auth_proxy_patch.yaml:13]: Good question, though duplicate - v0.8.0 is the latest available. We need someone drive them upload to latest.

File: controllers/addon_controller.go [hunk 20]
```diff
@@ -161,7 +168,7 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
 	// Process addon instance
 	ret, procErr := r.processAddon(ctx, log, instance, wfl)
 
-	// Always update cache, status
+	// Always update cache, status except errors
```
[the remark this answers is not in the document]
Comment (ccfishk) [reply controllers/addon_controller.go:171]: we do not want to update cache if there is an error. Sorry for the confusing, but it is the intention.

File: main.go (hunk 17 above)
Comment (ccfishk) [reply main.go:55]: This is k8s controller pattern : tells the client that, if any cache is used, to bypass it for the given objects. The controllers are specifically interested in those objects.

File: controllers/addon_controller.go (hunk 16 above)
Comment (ccfishk) [reply controllers/addon_controller.go:111]: You mean when there is 1000s workflows in the cluster ?

Comment (ccfishk) [reply controllers/addon_controller.go]: yes, the load tests with 2k is pretty positive. still collecting data.

File: config/default/manager_auth_proxy_patch.yaml (hunk 19 above)
[the remark this answers is not in the document]
Comment (figo) [reply config/default/manager_auth_proxy_patch.yaml:13]: I just curious what happened to the `0.10` image.

File: controllers/addon_controller.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (ccfishk) [reply controllers/addon_controller.go:170]: yea, wording can always be refined for clarity.

File: main.go (hunk 17 above)
Comment (ccfishk) [reply main.go:55]: Nice catch, explaination added.

File: controllers/addon_controller.go (hunk 16 above)
Comment (ccfishk) [reply controllers/addon_controller.go]: @kevdowney 
kubectl -n addon-manager-system get addons | wc -l
1208

Process                    CPU  Mem
addon-manager       11.7  0.8
workflow-controller 11.4  0.7

Comment (ccfishk) [reply controllers/addon_controller.go]: Uploaded load tests results.

File: controllers/addon_controller.go [hunk 21]
```diff
@@ -17,43 +17,52 @@ package controllers
...
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
 	addonmgrv1alpha1 "github.com/keikoproj/addon-manager/api/v1alpha1"
 	"github.com/keikoproj/addon-manager/pkg/addon"
+	AddonPkg "github.com/keikoproj/addon-manager/pkg/addon"
```
Comment (kevdowney) [inline controllers/addon_controller.go:52]:
```suggestion
	addonPkg "github.com/keikoproj/addon-manager/pkg/addon"
```

File: controllers/addon_controller.go [hunk 22]
```diff
@@ -173,71 +182,106 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
+
+	nsInformers := dynamicinformer.NewFilteredDynamicSharedInformerFactory(r.dynClient, time.Minute*30, workflowDeployedNS, nil)
+	wfInf := nsInformers.ForResource(common.WorkflowGVR())
+	// Watch workflows created by addon only in addon-manager-system namespace
+	if err := c.Watch(&source.Informer{Informer: wfInf.Informer().(cache.Informer)}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &addonmgrv1alpha1.Addon{},
+	}, predicate.NewPredicateFuncs(r.workflowHasMatchingNamespace)); err != nil {
+		return nil, err
 	}
```
Comment (kevdowney) [inline controllers/addon_controller.go:211]: We need to start the cache and sync for the informer.
```
err := mgr.Add(manager.RunnableFunc(func(s <-chan struct{}) error {
		nsInformers.Start(s)
		nsInformers.WaitForCacheSync(s)
		<-s
		return nil
	}))
```

File: controllers/addon_controller.go [hunk 23]
```diff
@@ -173,71 +182,106 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
+}
+
+func (r *AddonReconciler) filterAddons(obj client.Object) bool {
+	addon, ok := obj.(*addonmgrv1alpha1.Addon)
+	if !ok {
+		r.Log.Error(fmt.Errorf("unexpected object type in addon watch predicates"), "expected", "*addonmgrv1alpha1.Addon", "found", reflect.TypeOf(obj))
+		return false
+	}
+
+	if AddonPkg.IsDuplicate(addon, r.versionCache) {
```
Comment (kevdowney) [inline controllers/addon_controller.go:251]: Why is this here now?

File: controllers/addon_controller.go (hunk 22 above)
Comment (ccfishk) [reply controllers/addon_controller.go:211]: Yes, the newer version is needed.

Comment (ccfishk) [reply controllers/addon_controller.go:211]: done
86a35bc014041f67f38c80fe45919e37880d6661

File: controllers/addon_controller.go (hunk 23 above)
Comment (ccfishk) [reply controllers/addon_controller.go]: @kevdowney we discussed this previously, following is the discussion
https://github.com/keikoproj/addon-manager/pull/96#discussion_r803175059

File: controllers/addon_controller.go [hunk 24]
```diff
@@ -173,71 +182,112 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
-	}))
+	// Watch workflows created by addon only in addon-manager-system namespace
+	if err := c.Watch(&source.Informer{Informer: wfInf.Informer().(cache.Informer)}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &addonmgrv1alpha1.Addon{},
+	}, predicate.NewPredicateFuncs(r.workflowHasMatchingNamespace)); err != nil {
+		return nil, err
+	}
+
+	wfInforms := NewwfInformers(generatedInformers, nsInformers, stopChan)
```
Comment (kevdowney) [inline controllers/addon_controller.go:212]:
```suggestion
	wfInforms := NewWfInformers(generatedInformers, nsInformers, stopChan)
```

Comment (ccfishk) [reply controllers/addon_controller.go]: 8dc2cd3659bf433769c6d5d10180413b08acea85

File: controllers/addon_controller.go [hunk 25]
```diff
@@ -173,71 +182,112 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
+	err = mgr.Add(wfInforms)
```
Comment (kevdowney) [inline controllers/addon_controller.go:213]: This is supposed to be a lambda function, RunnableFunc() that should do some action not return, unless that's changed from the controller-runtime version upgrade... How is passing the informer constructor performing the the `wfInforms.Start()` method?

Comment (kevdowney) [reply controllers/addon_controller.go:213]: Also, `generatedInformers` was never initialized and you've removed usage for that so I don't think that would need to be passed to the constructor.

Comment (ccfishk) [reply controllers/addon_controller.go:214]: wsInforms.Start() is a manager.Runnable, in the newer controller-runtime, manager.Add() requires Runnable(context) instead of RunnableFunc().

File: controllers/addon_controller.go [hunk 26]
```diff
@@ -173,71 +181,113 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
-			OwnerType:    &addonmgrv1alpha1.Addon{},
-		})
+	if err := c.Watch(&source.Informer{Informer: wfInf.Informer()}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &addonmgrv1alpha1.Addon{},
+	}, predicate.NewPredicateFuncs(r.workflowHasMatchingNamespace)); err != nil {
+		return nil, err
+	}
 
 	generatedInformers = informers.NewSharedInformerFactory(r.generatedClient, time.Minute*30)
```
Comment (kevdowney) [inline controllers/addon_controller.go:211]: We still don't use this, just get rid of all mentions of `generatedInformers` and we should be good.

Comment (ccfishk) [reply controllers/addon_controller.go:211]: 985848b71f6f4a3888cd44aa78cd82ba264bf0f0'dib
Done !

File: controllers/addon_controller.go [hunk 27]
```diff
@@ -173,71 +181,112 @@ func (r *AddonReconciler) execAddon(ctx context.Context, req reconcile.Request,
...
+}
+
+func (r *AddonReconciler) filterAddons(obj client.Object) bool {
+	addon, ok := obj.(*addonmgrv1alpha1.Addon)
+	if !ok {
+		r.Log.Error(fmt.Errorf("unexpected object type in addon watch predicates"), "expected", "*addonmgrv1alpha1.Addon", "found", reflect.TypeOf(obj))
+		return false
+	}
+
+	if addonPkg.IsDuplicate(addon, r.versionCache) {
```
Comment (kevdowney) [inline controllers/addon_controller.go:256]: Get rid of this!

File: controllers/suite_test.go [hunk 28]
```diff
@@ -86,9 +87,6 @@ var _ = BeforeSuite(func(done Done) {
 	Expect(err).ToNot(HaveOccurred())
 	Expect(mgr).ToNot(BeNil())
 
-	err = NewAddonReconciler(mgr, ctrl.Log.WithName("controllers").WithName("Addon")).SetupWithManager(mgr)
```
Comment (kevdowney) [inline controllers/suite_test.go:89]: Why are these lines removed, we still want to test setup? We

Comment (ccfishk) [reply controllers/suite_test.go:89]: Because we moved new way to add controller with manager.

**Review by kevdowney — APPROVED** [review]
LGTM. Let’s evaluate after merge if we need more code coverage.
